client/src/app/auth/auth.service.ts

```diff
@@ -1,66 +1,183 @@
-      if (profile) {
-        self.userProfile = profile;
-      }
-      cb(err, profile);
+  authorize(): void {
+    this.getOpenIdConfig().subscribe((openIdConfig) => {
+      const authorizeEndpoint = openIdConfig.authorization_endpoint;
+      const params = {
+        redirect_uri: this.config.callbackUrl,
+      };
+
+      const urlParams: string = this.buildSearchParams(params);
+      const url: string = `${authorizeEndpoint}?${urlParams}`;
+      this.redirect(url);
+    })
+  }
+
+  public isAuthenticated(): boolean {
+    // Check whether the current time is past the
+    // access token's expiry time
+    const expiresAt = JSON.parse(localStorage.getItem('expires_at'));
+    return new Date().getTime() < expiresAt;
+  }
+
+  authorizeSilently(): void {
+    this.getOpenIdConfig().subscribe((openIdConfig) => {
+      const authorizeEndpoint = openIdConfig.authorization_endpoint
+      const params = {
+        scope: 'openid profile read:profile email',
+        redirect_uri: this.config.callbackUrl,
+        prompt: 'none'
+      };
+
+      const urlParams: string = this.buildSearchParams(params);
+      const url: string = `${authorizeEndpoint}?${urlParams}`;
+
+      this.runIframe(url, `https://${this.config.domain}`).then((data) => {
+        console.debug('code result', data);
+        this.setSession(data);
+      }).catch((error) => {
+        console.error('Error in silent authentication', error);
+      })
+    })
+  }
+
+  setUpSilentAuth(): void {
+    this.silentAuthTimer = TimerObservable.create(this.silentAuthInterval, this.silentAuthInterval)
+      .takeWhile(() => this.isAuthenticated());
+
+    this.silentAuthSubscription = this.silentAuthTimer.subscribe(() => {
+      this.authorizeSilently();
     });
   }
 
-  public logout(): void {
+  logout(): void {
     // Remove tokens and expiry time from localStorage
     localStorage.removeItem('access_token');
     localStorage.removeItem('id_token');
     localStorage.removeItem('expires_at');
-    // Go back to the home route
+
+    this.getOpenIdConfig().pipe(mergeMap((openIdConfig) => {
+      const logoutEndpoint = openIdConfig.logout_endpoint;
+      return this.http.get(logoutEndpoint);
+    })).subscribe(
+      ()=> {
+      console.log('successfully logged out...');
+    }, err => {
+      console.error('Error in logging out', err)
+    });
+
     this.router.navigate(['/login']);
   }
 
-  public isAuthenticated(): boolean {
-    // Check whether the current time is past the
-    // access token's expiry time
-    const expiresAt = JSON.parse(localStorage.getItem('expires_at'));
-    return new Date().getTime() < expiresAt;
+  setSession(params): void {
+    const accessToken = params['access_token'];
+    const expiresIn = parseInt(params['expires_in'], 10)
+    const idToken = params['id_token'];
+    const expiresAt = JSON.stringify((expiresIn * 1000) + new Date().getTime());
+    localStorage.setItem('access_token', accessToken);
```

Comment:
If we store the token in localStorage, isn't it susceptible to cross-site scripting attacks, which was the motivation for Silent Authentication in the first place?

Reply:
the motivation for silent authentication was to avoid providing a refresh token to the browser. Only an access token with a short lifespan is provided to the browser.